From 0bbeec07c51c427a6d42073ab62f19bdd4b06408 Mon Sep 17 00:00:00 2001
From: Alexandre Julliard <julliard@winehq.org>
Date: Mon, 7 Apr 2025 15:59:22 +0200
Subject: [PATCH 06/48] ntdll: Add a helper function to open the main image as
 .so file.

(cherry picked from commit a0f42351c7cd0c95eb4a80aca55e1ce0ea6bbdcc)

CW-Bug-Id: #26492
---
 dlls/ntdll/unix/loader.c | 38 ++++++++++++++++++++++++++------------
 1 file changed, 26 insertions(+), 12 deletions(-)

diff --git a/dlls/ntdll/unix/loader.c b/dlls/ntdll/unix/loader.c
index 45b24fe155b..3006b7ad874 100644
--- a/dlls/ntdll/unix/loader.c
+++ b/dlls/ntdll/unix/loader.c
@@ -1556,6 +1556,31 @@ static NTSTATUS open_builtin_so_file( const char *name, OBJECT_ATTRIBUTES *attr,
 }
 
 
+/***********************************************************************
+ *           open_main_image_so_file
+ */
+static NTSTATUS open_main_image_so_file( const char *name, UNICODE_STRING *nt_name, void **module,
+                                         SECTION_IMAGE_INFORMATION *image_info )
+{
+    struct pe_image_info pe_info;
+    NTSTATUS status;
+
+    /* remove .so extension from Windows name */
+    if (nt_name->Length > 3 * sizeof(WCHAR))
+    {
+        static const WCHAR soW[] = {'.','s','o',0};
+        WCHAR *p = nt_name->Buffer + nt_name->Length / sizeof(WCHAR);
+        if (!wcsicmp( p - 3, soW ))
+        {
+            p[-3] = 0;
+            nt_name->Length -= 3 * sizeof(WCHAR);
+        }
+    }
+    status = dlopen_dll( name, nt_name, module, &pe_info, FALSE );
+    if (!status) virtual_fill_image_information( &pe_info, image_info );
+    return status;
+}
+
 /***********************************************************************
  *           find_builtin_dll
  */
@@ -1787,15 +1812,12 @@ BOOL is_builtin_path( const UNICODE_STRING *path, WORD *machine )
 static NTSTATUS open_main_image( WCHAR *image, void **module, SECTION_IMAGE_INFORMATION *info,
                                  enum loadorder loadorder, USHORT machine )
 {
-    static const WCHAR soW[] = {'.','s','o',0};
     UNICODE_STRING nt_name;
     OBJECT_ATTRIBUTES attr;
-    struct pe_image_info pe_info;
     SIZE_T size = 0;
     char *unix_name;
     NTSTATUS status;
     HANDLE mapping;
-    WCHAR *p;
 
     if (loadorder == LO_DISABLED) NtTerminateProcess( GetCurrentProcess(), STATUS_DLL_NOT_FOUND );
 
@@ -1816,15 +1838,7 @@ static NTSTATUS open_main_image( WCHAR *image, void **module, SECTION_IMAGE_INFO
     }
     else if (status == STATUS_INVALID_IMAGE_NOT_MZ && loadorder != LO_NATIVE)
     {
-        /* remove .so extension from Windows name */
-        p = image + wcslen(image);
-        if (p - image > 3 && !wcsicmp( p - 3, soW ))
-        {
-            p[-3] = 0;
-            nt_name.Length -= 3 * sizeof(WCHAR);
-        }
-        status = dlopen_dll( unix_name, &nt_name, module, &pe_info, FALSE );
-        if (!status) virtual_fill_image_information( &pe_info, info );
+        status = open_main_image_so_file( unix_name, &nt_name, module, info );
     }
     free( unix_name );
     return status;
-- 
2.53.0

